### PR TITLE
Release with rig

### PR DESCRIPTION
**Pull request by rossabaker**
- Backport #1426 to appease Appveyor
- Moves a bunch of yaml and shell into Scala
- `ci`: a universal build target, even locally
- Generates versions.md from tags
- Publishes signed builds, even for snapshots
- If build is not a snapshot, will publish to Sonatype. [untested]

### CODE REVIEW

File: project/Http4sPlugin.scala [hunk 1]
```diff
@@ -53,7 +87,117 @@ object Http4sPlugin extends AutoPlugin {
...
+      val buildData: String =
+        s"""
+           |[versions]
+           |"http4s.api" = "$major.$minor"
+           |"http4s.current" = "${version.value}"
+           |"http4s.doc" = "${docExampleVersion(version.value)}"
+           |scalaz = "${scalazVersion.value}"
+           |circe = "${circeJawn.revision}"
+           |cryptobits = "${cryptobits.revision}"
+           |"argonaut-shapeless_6.2" = "1.2.0-M5"
```
Comment (ChristopherDavenport) [inline project/Http4sPlugin.scala]: Why is this hard-coded?

File: project/Http4sPlugin.scala [hunk 2]
```diff
@@ -102,6 +246,27 @@ object Http4sPlugin extends AutoPlugin {
...
+    JGit(file).tags.collect {
+      case ref if ref.getName.startsWith("refs/tags/v") =>
+        Version(ref.getName.substring("refs/tags/v".size))
+    }.foldLeft(Map.empty[(Int, Int), Version]) {
+      case (m, Some(v)) =>
+        def toMinor(v: Version) = (v.major, v.subversions.headOption.getOrElse(0))
+        def patch(v: Version) = v.subversions.drop(1).headOption.getOrElse(0)
+        def milestone(v: Version) = v.qualifier match {
+          case Some(q) if q.startsWith("-M") => q.substring(2).toInt
+          case Some(q) if q.startsWith("-RC") => q.substring(3).toInt + 1000000
```
Comment (ChristopherDavenport) [inline project/Http4sPlugin.scala]: I may be missing it, whats with the `+ 1000000`

File: project/Http4sPlugin.scala [hunk 3]
```diff
@@ -53,7 +87,117 @@ object Http4sPlugin extends AutoPlugin {
...
+          checkSnapshotDependencies,
+          inquireVersions,
+          setReleaseVersion,
+          tagRelease
+        )
+      else
+        Seq.empty[ReleaseStep],
+
+      Seq[ReleaseStep](
+        runTestWithCoverage,
```
Comment (ChristopherDavenport) [inline project/Http4sPlugin.scala]: You mentioned in gitter this broke coverage, why isn't this running that?

File: project/Http4sPlugin.scala [hunk 4]
```diff
@@ -53,7 +87,117 @@ object Http4sPlugin extends AutoPlugin {
...
+    }
+  ) ++ releaseSettings
+
+  val releaseSettings = Seq(
+    // Reset a couple sbt-release defaults that rig changed
+    releaseVersion := { ver => Version(ver).map(_.withoutQualifier.string).getOrElse(versionFormatError) },
+    releaseTagName := s"v${if (releaseUseGlobalVersion.value) (version in ThisBuild).value else version.value}",
+
+    releaseProcess := Seq(
+      if (!isSnapshot.value)
```
Comment (ChristopherDavenport) [inline project/Http4sPlugin.scala]: Can these be broken out into individual functions for greater clarity? Functional programmer in me looks at this as practically imperative.

File: project/Http4sPlugin.scala (hunk 1 above)
Comment (rossabaker) [reply project/Http4sPlugin.scala]: I think the main build doesn't depend on it, but it also doesn't depend on cryptobits.  This is not new, but it is curious.

File: project/Http4sPlugin.scala (hunk 4 above)
Comment (rossabaker) [reply project/Http4sPlugin.scala]: I tried earlier but struggled a bit with names.  I'll give it another go.

File: project/Http4sPlugin.scala (hunk 2 above)
Comment (rossabaker) [reply project/Http4sPlugin.scala]: I wanted to sort RCs after milestones.  I should turn this into a tuple and Do It Right.

File: project/Http4sPlugin.scala (hunk 3 above)
Comment (rossabaker) [reply project/Http4sPlugin.scala]: I've said a lot of things.  I _think_ it is running now?

File: project/Http4sPlugin.scala (hunk 4 above)
Comment (rossabaker) [reply project/Http4sPlugin.scala]: I'm not finding it much clearer when I give them names.

File: project/Http4sPlugin.scala (hunk 1 above)
Comment (rossabaker) [reply project/Http4sPlugin.scala]: This one is not actually compiled into the tuts.  Unless it ever is, adding it as a dependency would just download another piece of the internet.

File: project/Http4sPlugin.scala (hunk 4 above)
Comment (rossabaker) [reply project/Http4sPlugin.scala]: See if you like 0fc93f6 better.

**Comment by ChristopherDavenport** [comment]
Caught again by [timeout spec](https://travis-ci.org/http4s/http4s/jobs/281386589#L3646)

I'd recommend to rerun that build, to see if it can pass. As this is such a crucial PR I'd prefer it pass CI.

**Comment by rossabaker** [comment]
The fixes there haven't been backported from 0.18 yet.  I'll restart.

I don't know why appveyor doesn't like this.

**Comment by rossabaker** [comment]
Cool, I guessed right on Appveyor, too!

**Pull request merged**